File: content/problems/0238-product-of-array-except-self/solutions/prefix-suffix/Solution.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] answer = new int[nums.length]; answer[0] = 1;
        for (int index = 1; index < nums.length; index++) answer[index] = answer[index - 1] * nums[index - 1];
        int suffix = 1;
        for (int index = nums.length - 1; index >= 0; index--) { answer[index] *= suffix; suffix *= nums[index]; }
        return answer;
    }
}

File: content/problems/0238-product-of-array-except-self/solutions/prefix-suffix/Solution.kt
class Solution {
    fun productExceptSelf(nums: IntArray): IntArray {
        val answer = IntArray(nums.size); answer[0] = 1
        for (index in 1 until nums.size) answer[index] = answer[index - 1] * nums[index - 1]
        var suffix = 1
        for (index in nums.lastIndex downTo 0) { answer[index] *= suffix; suffix *= nums[index] }
        return answer
    }
}
